apm-sniffer/apm-sdk-plugin/undertow-plugins/undertow-2.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/undertow/v2x/RoutingHandlerInterceptor.java

```diff
@@ -0,0 +1,96 @@
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceMethodsAroundInterceptor;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.MethodInterceptResult;
+import org.apache.skywalking.apm.plugin.undertow.v2x.util.TraceContextUtils;
+
+import java.lang.reflect.Method;
+
+/**
+ * @author AI
+ * 2019-07-25
+ */
+public class RoutingHandlerInterceptor implements InstanceMethodsAroundInterceptor {
+
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, MethodInterceptResult result) throws Throwable {
+        int httpHandlerIndex = -1;
+        for (int index = 0; index < allArguments.length; index++) {
+            Object any = allArguments[index];
+            if (any instanceof HttpHandler) {
+                httpHandlerIndex = index;
+                break;
+            }
+        }
+        if (httpHandlerIndex > -1) {
+            HttpHandler handler = (HttpHandler) allArguments[httpHandlerIndex];
+            String template = (String) allArguments[1];
+            allArguments[httpHandlerIndex] = new TracingHandler(template, handler);
+            TraceContextUtils.enabledRoutingHandlerTracing();
+        }
+    }
+
+    @Override
+    public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, Object ret) throws Throwable {
+        return ret;
+    }
+
+    @Override
+    public void handleMethodException(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, Throwable t) {
+        ContextManager.activeSpan().errorOccurred().log(t);
+    }
+
+    private static class TracingHandler implements HttpHandler {
+        private final String template;
+        private final HttpHandler next;
+
+        TracingHandler(String template, HttpHandler handler) {
+            this.next = handler;
+            this.template = template;
+        }
+
+        @Override
+        public void handleRequest(HttpServerExchange exchange) throws Exception {
+            final AbstractSpan span = TraceContextUtils.buildUndertowEntrySpan(exchange, template);
+            exchange.addExchangeCompleteListener(new ExchangeCompletionListener() {
+                @Override
+                public void exchangeEvent(HttpServerExchange httpServerExchange, NextListener nextListener) {
+                    if (httpServerExchange.getStatusCode() >= 400) {
+                        span.errorOccurred();
+                        Tags.STATUS_CODE.set(span, Integer.toString(httpServerExchange.getStatusCode()));
+                    }
+                    span.asyncFinish();
+                    nextListener.proceed();
+                }
+            });
+            next.handleRequest(exchange);
+            span.prepareForAsync();
```

Comment:
`span.prepareForAsync();` should call before `next.handleRequest(exchange)` because in some low latency request, when `span.asyncFinish();` execute the `prepareForAsync` may not execute which may cause error.